apps/opencs/view/render/terrainshapemode.hpp

```diff
@@ -0,0 +1,172 @@
+
+            /// Get altered height values around one vertex
+            void updateKeyHeightValues(const CSMWorld::CellCoordinates& cellCoords, int inCellX, int inCellY, float* thisHeight,
+                float* thisAlteredHeight, float* leftHeight, float* leftAlteredHeight, float* upHeight, float* upAlteredHeight,
+                float* rightHeight, float* rightAlteredHeight, float* downHeight, float* downAlteredHeight);
+
+            ///Limit steepness based on either X or Y and return false if steepness is limited
+            void compareAndLimit(const CSMWorld::CellCoordinates& cellCoords, int inCellX, int inCellY, float* limitedAlteredHeightXAxis,
+                float* limitedAlteredHeightYAxis, bool* steepnessIsWithinLimits);
+
+            /// Check that the edit doesn't break save format limits, fix if necessary, return true if slope steepness is within limits
+            bool limitAlteredHeights(const CSMWorld::CellCoordinates& cellCoords, bool reverseMode = false);
+
+            /// Handle brush mechanics for terrain shape selection
+            void selectTerrainShapes (const std::pair<int, int>& vertexCoords, unsigned char selectMode, bool dragOperation);
+
+            /// Push terrain shape edits to command macro
+            void pushEditToCommand (const CSMWorld::LandHeightsColumn::DataType& newLandGrid, CSMDoc::Document& document,
+                CSMWorld::IdTable& landTable, std::string cellId);
+
+            /// Push land normals edits to command macro
+            void pushNormalsEditToCommand(const CSMWorld::LandNormalsColumn::DataType& newLandGrid, CSMDoc::Document& document,
+                CSMWorld::IdTable& landTable, std::string cellId);
+
+            /// Generate new land map LOD
+            void pushLodToCommand(const CSMWorld::LandMapLodColumn::DataType& newLandMapLod, CSMDoc::Document& document,
+                CSMWorld::IdTable& landTable, std::string cellId);
+
+            /// Create new cell and land if needed
+            bool allowLandShapeEditing(std::string textureFileName);
+
+        private:
+            std::string mCellId;
+            std::string mBrushTexture;
+            int mBrushSize;
+            int mBrushShape;
+            std::vector<std::pair<int, int>> mCustomBrushShape;
+            CSVWidget::SceneToolShapeBrush *mShapeBrushScenetool;
+            int mDragMode;
+            osg::Group* mParentNode;
+            bool mIsEditing;
+            std::unique_ptr<TerrainSelection> mTerrainShapeSelection;
+            int mTotalDiffY;
+            std::vector<CSMWorld::CellCoordinates> mAlteredCells;
+            osg::Vec3d mEditingPos;
+            int mShapeEditTool;
+            int mShapeEditToolStrength;
+            int mTargetHeight;
+
+            PagedWorldspaceWidget& getPagedWorldspaceWidget();
+
+        signals:
+            void passBrushTexture(std::string brushTexture);
```

Comment:
Pass strings as references, not copies (this is not the only instance of that).

Reply:
I fixed the others at 8ee6164 , but this one is a Qt signal, so I'm not sure if it's useful, or even if it may bring some problems at some platforms. There was some talk at Stackoverflow that they'd be reverted back to copies anyway.

Reply:
Link? In Qt documentation, there's one place where they pass a QString reference to a signal.

Reply:
https://stackoverflow.com/questions/10405739/const-ref-when-sending-signals-in-qt

I really have no other knowledge on the subject, it might or might not be better.

Reply:
I dunno. The answers imply it's better to pass a reference than not because if it's passed as a reference, it can end up both being copied and being passed as a reference, and if it's passed as a copy it will always be copied.

Reply:
Gave this one more look, and was about to change it to const ref, as it should also work. However, this specific signal is passBrush**Texture**, totally unused and removable in the first place... duh..